tests/integration_tests/test_public_api.py

```diff
@@ -126,21 +114,25 @@ def test_public_api(cloud: BaseCloud):
         # Not sure there's a great way to test this other than not raising
         cloud.image_serial(image_id)
 
-    with launch_instance(
-        cloud, image_id=image_id, user_data=cloud_config, wait=False
-    ) as instance:
-        instance.wait()
-        exercise_instance(instance)
+    try:
+        with cloud.launch(
+            image_id=image_id, user_data=cloud_config
+        ) as instance:
+            instance.wait()
+            exercise_instance(instance)
 
-        instance.clean()
-        result = instance.execute("sudo rm /var/tmp/example.txt")
-        snapshot_id = cloud.snapshot(instance)
+            instance.clean()
+            instance.execute("sudo rm /var/tmp/example.txt")
+            snapshot_id = cloud.snapshot(instance)
 
-    try:
-        with launch_instance(
-            cloud, image_id=snapshot_id, user_data=cloud_config, wait=False
-        ) as instance_from_snapshot:
-            instance_from_snapshot.wait()
-            exercise_instance(instance_from_snapshot)
+        try:
+            with cloud.launch(
+                image_id=snapshot_id, user_data=cloud_config
+            ) as instance_from_snapshot:
+                instance_from_snapshot.wait()
+                exercise_instance(instance_from_snapshot)
+        finally:
+            cloud.delete_image(snapshot_id)
     finally:
-        cloud.delete_image(snapshot_id)
+        if isinstance(cloud, pycloudlib.Azure):
+            cloud.delete_resource_group()
```

Comment:
This and the `finally:  cloud.delete_image(snapshot_id)` above could be automatically removed if our BaseCloud were used a context manager that collected any resources created during that test run and purged those resources on `__exit__` from the context.  We'd probably want to also create a public method `BaseCloud.release_resources`  or `delete_resources` or something to allow folks not using a context manager to manually invoke this at the end of any tests they are running anyway.